doc/console/SONiC-Console-Switch-High-Level-Design.md

```diff
@@ -0,0 +1,488 @@
+    + [4.5.1 Basic Usage](#451-basic-usage)
+    + [4.5.2 Port based Forwarding](#452-port-based-forwarding)
+    + [4.5.3 IP based Forwarding](#453-ip-based-forwarding)
+- [5 Error Handling](#5-error-handling)
+- [6 Serviceability and Debug](#6-serviceability-and-debug)
+- [7 Warm Boot Support](#7-warm-boot-support)
+- [8 Scalability](#8-scalability)
+
+# List of Tables
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+# Revision
+| Rev |     Date    |          Authors             | Change Description                |
+|:---:|:-----------:|:----------------------------:|-----------------------------------|
+| 0.1 | 08/28/2020  |  Jing Kan | Initial version                   |
+
+# About this Manual
+
+This document provides general information about Console switch features implementaion in SONiC
+
+# Scope
+
+This document describes the functionality and high level design of the Console switch features in SONiC.
+
+# Definition/Abbreviation
+
+### Table 1: Abbreviations
+| Term | Meaning
+|---|---|
+| SSH | Secure Shell |
+| CLI | Command Line Interface |
+| OS | Operating System |
+| USB | Universal Serial Bus |
+
+# 1 Feature Overview
+Console Switch enable user to manage routers/switchs via console port.
+
+## 1.1 Requirements
+
+### 1.1.1 Functional Requirements
+
+1. Support configuration of flow control on each console port
+1. Support configuration of baudrate on each console port
+1. Support access console line via reverse SSH
+
+### 1.1.2 Configuration and Management Requirements
+
+1. Support configuration using SONiC CONFIG_DB
+1. Support configuration using SONiC MINIGRAPH
+
+## 1.2 Design Overview
+
+### 1.2.1 Basic Approach
+
+Enable SONiC device to manage connected Console device(Serial Hub).
+
+Persist console configurations to control how to view/connect to a device via serial link, includes:
+- Driver
+- Line Number
+- Flow Control switch
+- Remote Device Name
+- Management IP (optional)
+
+SONiC OS provide a unified cli utility to help user connect and manage remote device.
+User use SSH to connect to SONiC device via SONiC management IP.
+
+### 1.2.2 Container
+
+No new containers are introduced.
+
+# 2 Functionality
+
+Refer to section 1
+
+## 2.1 Target Deployment Use Cases
+
+Configure/Manage remote devices via console port in SONiC.
+
+## 2.2 Functional Description
+
+Refer to section 1.1
+
+## 2.3 Limitations
+
+- Configuration is only supported on console port.
+
+# 3 Design
+
+## 3.1 Overview
+
+The design overview at a high level:
+
+![SONiC Console Topo](./ConsoleTopo.png)
+
+SONiC OS is run on a regular network router or switch.
+
+The Console device is serial hub which enable user to manage multiple network devices via console.
+
+Console device(s) are connect to the SONiC device as add-on. In current design, there are no limitation to the way you can connect and no limit to the number of add-on devices. All limitation is on SONiC switch's hardware interfaces and console device's driver.
```

Comment:
The HLD mostly focuses on a design where the console controllers are connected as a USB device or some other kind of external device.
I just want to confirm that the case where the console controller are directly integrated on the switch and managed by kernel drivers in SONiC OS is supported.

You could think of a switch that has a mix of Ethernet and Console front panel ports and manage them all within the same system without any external device and exposing all these consoles as tty in SONiC OS. Would that work as well ?

```
+-------------------------+       +------------------------+
|  Switch                 |       |  Device 1              |
|  +-------------------+  |       |  +--------+ +--------+ |
|  |SONiC OS           |  |   +------>Console | |Ethernet| |
|  |                   |  |   |   |  +--------+ +--------+ |
|  +-------------------+  |   |   |                        |
|                         |   |   +------------------------+
|  +--------+ +--------+  |   |
|  |Ethernet| |Console1+------+   +------------------------+
|  +--------+ +--------+  |       |  Device 2              |
|  +--------+ +--------+  |       |  +--------+ +--------+ |
|  |Console | |Console2+------------->Console | |Ethernet| |
|  +--------+ +--------+  |       |  +--------+ +--------+ |
|                         |       |                        |
+-------------------------+       +------------------------+
```

Reply:
This is a good question!

This HLD is focus on support external serial hub.
In my view, the top consideration is extensibility: not all serial hub can install a OS but we can still using them.

In other hand, we can also support embedded console ports by exposing them to tty with some adapter, then we can use same flow to control remote devices, what's more, if we don't want to expose tty, we can improve the consult in future to use SAI to manage embedded console ports.

@yxieca  could you comments on this? Do we have plan to support embedded console without exposing tty?

Reply:
I don't necessary see the relevance of the serial hub being USB in the design.
A known USB hub will be automatically detected by the kernel and its driver will allocate a tty for each serial port available.
The same applies to serial ports seated directly on the switch and managed by SONiC OS.
A kernel driver will expose them through `/dev/tty%s%d`

I see in your design that the `CONSOLE_PORT_TABLE` table of `CONFIG_DB` seems to support that usecase.
Are the fields in the port table meant to be used this way `"/dev/tty%s%d" % (driver, row_name)` ?

I would suggest adding a field `id` since this model would not work if there were different drivers in use.
Each driver gets its own set of ids which means that it's possible to have similar values.
For example `/dev/ttyACM0` and `/dev/ttyUSB0`.
The key could alternatively be `"USB1"` instead of just `"1"`. Or even `"USB|1"`.

The reason I highlight this is that it would prevent a switch having internal tty ports from having usb tty ports.
You would have to make a choice to avoid key conflict in `CONSOLE_PORT_TABLE`.

Let me know what you think.

Reply:
Yes, actually, the USB topo is just a sample.

Your concern is make sense, that's why we introduce the field "Driver" in CONFIG_DB.

We hoping we can identify a console tty by using **(driver, port number)** which means we can't simply use the original device name here, otherwise the conflict might happens just like you mentioned.

In current design, the ```port number``` is the unique id, we will aliasing the original tty device name to a specified format to avoid device name conflict or port shifting. Here is another HLD regarding this: https://github.com/Azure/SONiC/pull/654/files

The ```driver``` type will help us to build a mapping between the physical port and specific device name.